wandb/sdk/wandb_run.py

```diff
@@ -1456,39 +1456,77 @@ def _datatypes_callback(self, fname: str) -> None:
-                value = row[k].get_config_value(k)
-                row[k] = row[k]._data
-                self._config_callback(val=value, key=key)
-            elif isinstance(row[k], CustomChart):
-                chart_keys.add(k)
-                key = row[k].get_config_key(k)
-                if row[k]._split_table:
-                    value = row[k].get_config_value(
-                        "Vega2", row[k].user_query(f"Custom Chart Tables/{k}_table")
+
+        def transform(key: Union[str, List[str]], value: Any) -> Any:
+            if isinstance(key, str):
+                key = [key]
+            if isinstance(value, dict):
+                return {k: transform([*key, k], v) for k, v in value.items()}
+            if isinstance(value, Visualize):
+                config_key = value.get_config_key(
+                    key if isinstance(key, str) else ".".join(key)
+                )
+                config_value = value.get_config_value(
+                    key if isinstance(key, str) else ".".join(key)
+                )
+                value = value._data
+                self._config_callback(val=config_value, key=config_key)
+                return value
+            elif isinstance(value, CustomChart):
+                formatted_key = ".".join(key)
+                chart_keys.add(tuple(key))
+
+                config_key = value.get_config_key(
+                    key if isinstance(key, str) else ".".join(key)
+                )
+                if value._split_table:
+                    config_value = value.get_config_value(
+                        "Vega2",
+                        value.user_query(f"Custom Chart Tables/{formatted_key}_table"),
                     )
-                    split_table_set.add(k)
+                    split_table_set.add(tuple(key))
                 else:
-                    value = row[k].get_config_value(
-                        "Vega2", row[k].user_query(f"{k}_table")
+                    config_value = value.get_config_value(
+                        "Vega2", value.user_query(f"{formatted_key}_table")
                     )
-                row[k] = row[k]._data
-                self._config_callback(val=value, key=key)
+                value = value._data
+                self._config_callback(val=config_value, key=config_key)
+                return value
+            else:
+                return value
+
+        for k, v in row.items():
+            row[k] = transform(k, v)
 
         for k in chart_keys:
-            # remove the chart key from the row
-            # TODO: is this really the right move? what if the user logs
-            #     a non-custom chart to this key?
+            klist = list(k)
+            value = self._get_nested_value(row, klist)
+
             if k in split_table_set:
-                row[f"Custom Chart Tables/{k}_table"] = row.pop(k)
+                self._set_nested_value(row, klist, value, split_table=True)
```

Comment:
it looks like there are a lot of updates uncovered by tests, is it right? if so, mind adding more tests?

Reply:
@dmitryduev sounds great, I'll definitely add more! Sorry, this pr and changes came out to be much bigger than expected + a bunch of troubles setting up the environment, that is why it's so messy 😢